### DISCUSSION TITLE
Magick command - Limit output size for PNG rendition

### DISCUSSION THREAD

**Discussion by MisterJpa** [ImageMagick/ImageMagick#4104] [Help]
Hi everyone,

I'm trying to write a command to match to following requirement from my customer : 
- Input will be a tiff file with no size or ratio restriction
- Output should be a png file
- Output should always be a square
- The maximum size of the output should be 1920x1920.
- Output background should be transparent (input will always be transparent)
- Output weight should be less than 6MB

This command 
`magick myTifFile.tif[0] -resize 1920x1920 -background transparent -gravity Center -extent 1920x1920 myPngRendition.square.1920.1920.png`

match the 5 first requirement but I'm failing at matching the last one. 
Is there any way to limit the output size for a png ? 
On this thread : https://stackoverflow.com/questions/6917219/imagemagick-scale-jpeg-image-with-a-maximum-file-size, there are limiting the size for a jpeg file but I didn't find the way to do it for a png.. 

Many thanks by advance. 
Regards,
MisterJpa

**Comment by snibgo** [comment]
What version of IM, on what platform?

What images do you have? Cartoon-like graphics, or photographs? PNG compression is not designed for photos.

> Is there any way to limit the output size for a png ?

No. Unlike JPEG, PNG does not have a simple lossy size/quality tradeoff. PNG is always lossless. 

If your IM is Q16, then you can save as 8-bit by inserting "-depth 8" before writing the output, to generally halve the size. "-strip" will remove unneeded metadata, but will also remove any embedded ICC profile which may be a bad idea.

External programs such as pngcrush may also help.

If that doesn't do what you want, you might simplify the image: reduce it to 256 colours, remove high-frequency detail, and so on.

**Reply by MisterJpa** [reply]
Hi !

Thanks for answering !

Here is the version : ImageMagick 7.0.11-10 Q16 x86_64 2021-04-29 https://imagemagick.org                                                                                                                              
The system is running on a centOS 7.4. 

I've try the "-depth 8" but it was not really successfull. 

The input image is a tiff file, 57Mo. It's a photographs of a product (a product packshot). Unfortunately, I'm not allowed to provide the image... :( 

However, here are some extra informations. If I run the same command removing the [0], I got those 3 files : 
![image](https://user-images.githubusercontent.com/9430842/130453841-438658c0-a6e7-4f13-91cb-1eeb95a2cc44.png)

Running the same command with [2] will provide a png file that weight 2.1Mo and that looks exactly the same as the one weighting 57Mo. 
The [1] is just a white square. 

I'm not really familiar with image treatment (not my job) but if I'm right, tiff can be built as a stack of several images. 
However, I don't think that I can expect the input image to always be built this way.. 


From what I understand in your message, there is no way to ensure that the output will weight less than 6Mo. As it seems to be a technical requirement for salesforce to ingest images, I'll ask the customer if receiving a jpeg with white background is an option or not.

**Comment by snibgo** [comment]
A tiff file can contain any number of images. They may be related, such as at different sizes (in pixels), or with and without transparency. Or they may be unrelated. Your customer should say which image is required.

If your input image is already just 8 bits/channel/pixel, then "-depth 8" won't reduce the output size further.

**Reply by MisterJpa** [reply]
Hi !

I'm sorry, I totally forgot to answer.
Yes, the image for not reducing because it was already an 8 bits image.

After discussion and explanation, the customer is ok to go with white background image, jpeg format.

So this topic is no longer needed.

Many thanks for your help and explanation !
MisterJpa
